### PR TITLE
Improved dataloading with multiple sources without resampling.

### PR DESCRIPTION
**Pull request by GeorgiosSmyrnis**
This PR improves the deterministic dataloading when multiple data sources are included. The current implementation works by using a manifest for each data source, and then for each epoch using that manifest to pick shards to train for this particular epoch, according to the ratio defined by the source weights.

This PR is currently a WIP, with the following tasks remaining:
- [x] Include automated tests.
- [x] Ensure deterministic training.
- [x] Compare with training with resampling.
- [x] Update README.

### CODE REVIEW

**Comment by GeorgiosSmyrnis** [comment]
Replacement vs No Replacement

Including a file for comparison between the sampling with and without replacement.

**Comment by GeorgiosSmyrnis** [comment]
Following discussion with @gabrielilharco, splitting of the shards is no longer done strictly across epochs, but rather across chunks. Each chunk has roughly the number of samples required for the epoch, so if the number of tokens is large enough then there will be no difference.

This change was done in order to fix the issue of having only a few shards (less than the number of workers) within an epoch. This alleviates this issue by assigning more shards to an epoch, if needed.

File: open_lm/file_utils.py
Comment (sagadre) [inline open_lm/file_utils.py:96]: can we define what a chunk is here with a comment? is it a shard? a group of shards?

File: open_lm/file_utils.py
Comment (sagadre) [inline open_lm/file_utils.py:130]: is it possible this will infinite loop? maybe add some error handling in that case? what happens in the case where epochs*num_train_tokens > number of tokens in the dataset?

File: open_lm/file_utils.py
Comment (GeorgiosSmyrnis) [reply open_lm/file_utils.py:96]: It's a group of shards.
I'll add a comment clarifying this.

File: open_lm/file_utils.py
Comment (GeorgiosSmyrnis) [reply open_lm/file_utils.py:130]: Good point - some error handling is needed here.

File: open_lm/main.py
Comment (sagadre) [inline open_lm/main.py:146]: maybe call this function `load_checkpoint`?

File: open_lm/params.py
Comment (sagadre) [inline open_lm/params.py:93]: nit: dataset manifest instead of metadata?

File: open_lm/main.py
Comment (GeorgiosSmyrnis) [reply open_lm/main.py:146]: I wanted to keep this separate since loading the chunk id only makes sense when loading without resampling (whereas the existing ```load_model``` handles the actual model loading. It might be clearer this way I believe.

File: open_lm/params.py
Comment (sagadre) [inline open_lm/params.py:94]: are multiple manifest files supported (as suggested by the `nargs="+"` flag)? looking at `get_metadata_file(...)` it looks like only 1 file can be passed

Comment (GeorgiosSmyrnis) [reply open_lm/params.py:94]: Multiple manifest files are supported - ```get_metadata_file``` is just called separately on each.

File: tests/test_dataset_deterministic.py
Comment (sagadre) [inline tests/test_dataset_deterministic.py:17]: any way we can create a `/tests/assets` folder and put a small amount of mock data there for testing rather than pointing to local paths?

File: open_lm/main.py
Comment (sagadre) [reply open_lm/main.py:146]: oh i see, u only load checkpoint to get the "data_chunk" field, makes sense then!

File: tests/test_dataset_deterministic.py
Comment (sagadre) [reply tests/test_dataset_deterministic.py:17]: for this also probably good if we can check in some of our val data as opposed to rpj or pile shards

Comment (GeorgiosSmyrnis) [reply tests/test_dataset_deterministic.py:17]: Added a ```/tests/assets``` with a few tarfiles.

Comment (GeorgiosSmyrnis) [reply tests/test_dataset_deterministic.py:17]: For now I just added some of our training shards, lmk if we want to change that.

File: open_lm/file_utils.py
Comment (GeorgiosSmyrnis) [reply open_lm/file_utils.py:96]: Updated with a comment on this.

File: open_lm/file_utils.py
Comment (GeorgiosSmyrnis) [reply open_lm/file_utils.py:144]: Added some additional error handling, and found a corner case.
If you run out of shards and have to loop back, it is kind of problematic with the current string notation.

I updated the notation to list all shards e.g. ```pipe aws s3 cp s3://shards/shard_{00000, 00002, 00004}.tar -```, which I think has the desired effect of allowing arbitrary shards (not just continuous ones).

File: open_lm/params.py
Comment (GeorgiosSmyrnis) [reply open_lm/params.py:93]: Changed this into dataset-manifest.

**Comment by GeorgiosSmyrnis** [comment]
In the above implementation, there is a corner case where if the final group of shards does not have enough samples to cover a single epoch, it is discarded. This will matter if we exhaust the dataset, so I believe it should be handled as well.

File: open_lm/file_utils.py
Comment (achalddave) [inline open_lm/file_utils.py:148]: This is a pretty nontrivial function now, with a seemingly benign name. Could we add a docstring and/or possibly rename? It looks to me like the right docstring is:

```
"""
Collect shards to train on for this epoch, along with metadata.

This function looks through the list of all shards and returns a subset of shards
that result in >= num_samples samples and >= min_shards_needed shards. Note that
we cannot simply compute samples based on the number of shards, since shards can
have varying numbers of samples.
"""
```

and rename to `_get_shards_for_epoch`?

File: open_lm/file_utils.py
Comment (achalddave) [inline open_lm/file_utils.py:156]: I'm a little confused here; why do we need a loop here if get_shards_for_chunk also has a loop that packs shards until we have >= needed_samples_per_epoch? Is it only for the the last chunk, where we may have fewer samples than needed?

If so, I would have thought we should handle this in get_shards_for_chunk, but I suppose it's possible that we have one data source with not enough samples, and another with more than enough samples, and things work out for us serendipitously. If my understanding is correct, I think no changes needed, but might be good to drop a comment to this effect.

**Review by achalddave — APPROVED** [review]

**Comment by achalddave** [comment]
Approving because I think we've tested and iterated on this a bit, but left a few comments that would be nice-to-haves in this PR or next. I believe @Vaishaal is also planning to take a look, though, and we can wait for that before merge.

**Pull request merged**
